packages/coordinator/ts/deployer/deployer.service.ts

```diff
@@ -0,0 +1,943 @@
+                abi: VkRegistryFactory.abi,
+                functionName: "setVerifyingKeysBatch",
+                args: [
+                  config.VkRegistry.args.stateTreeDepth,
+                  config.VkRegistry.args.intStateTreeDepth,
+                  config.VkRegistry.args.messageTreeDepth,
+                  config.VkRegistry.args.voteOptionTreeDepth,
+                  MESSAGE_TREE_ARITY ** config.VkRegistry.args.messageBatchDepth,
+                  [EMode.QV, EMode.NON_QV],
+                  // @ts-expect-error - the abi has a more complex type for the processZkeys and tallyZkeys
+                  processZkeys,
+                  // @ts-expect-error - the abi has a more complex type for the processZkeys and tallyZkeys
+                  tallyZkeys,
+                ],
+              }),
+            }),
+          },
+        });
+        // set vKeys on the vk registry
+        const callGasLimitMultiplier = (opEstimate.callGasLimit / 100n) * 5n;
+
+        const userOpHashVkRegistry = await kernelClient.sendUserOperation({
+          userOperation: {
+            callGasLimit:
+              opEstimate.callGasLimit + callGasLimitMultiplier < 28000000n
+                ? opEstimate.callGasLimit + callGasLimitMultiplier
+                : 28000000n,
+            sender: kernelClient.account.address,
+            maxFeePerGas: gasEstimates.maxFeePerGas,
+            maxPriorityFeePerGas: gasEstimates.maxPriorityFeePerGas,
+            callData: await kernelClient.account.encodeCallData({
+              to: vkRegistryAddress as Hex,
+              value: 0n,
+              data: encodeFunctionData({
+                abi: VkRegistryFactory.abi,
+                functionName: "setVerifyingKeysBatch",
+                args: [
+                  config.VkRegistry.args.stateTreeDepth,
+                  config.VkRegistry.args.intStateTreeDepth,
+                  config.VkRegistry.args.messageTreeDepth,
+                  config.VkRegistry.args.voteOptionTreeDepth,
+                  MESSAGE_TREE_ARITY ** config.VkRegistry.args.messageBatchDepth,
+                  [EMode.QV, EMode.NON_QV],
+                  // @ts-expect-error - the abi has a more complex type for the processZkeys and tallyZkeys
+                  processZkeys,
+                  // @ts-expect-error - the abi has a more complex type for the processZkeys and tallyZkeys
+                  tallyZkeys,
+                ],
+              }),
+            }),
+          },
+        });
+
+        const receiptVkRegistry = await bundlerClient.waitForUserOperationReceipt({
+          hash: userOpHashVkRegistry,
+        });
+
+        if (!receiptVkRegistry.success) {
+          throw new Error(ErrorCodes.FAILED_TO_SET_VERIFYING_KEYS_ON_VK_REGISTRY);
+        }
+      }
+    } catch (error) {
+      this.logger.error("Failed to set verifying keys on vk registry: ", error);
+    }
+
+    // 5. maci (here we don't check whether one is already deployed, we just deploy it)
```

Comment:
just curious, why "don't check whether one is already deployed"?

Reply:
ok then, to check if it's deployed or not should be another function.

Reply:
yup please could you start writing down public endpoints you need on the frontend for the flow?